### ISSUE TITLE
Tree-shaking behaving differently after migration from 15.1.1 to 15.4.0

### ISSUE THREAD

**Issue by smithpb** [nrwl/nx#14483]
### Current Behavior

After migrating, the bundle size of our Angular application increased noticeably. Looking through the source map revealed that unused services were not being tree-shaken during the build process.

Source map before migration
<img width="1068" alt="Pre Migration" src="https://user-images.githubusercontent.com/41526919/213302391-83fc5e35-6c1e-48eb-b672-8a66f975c2dd.png">

Source map after migration
<img width="1003" alt="Post Migration" src="https://user-images.githubusercontent.com/41526919/213302527-9c33008d-d32e-4cf9-94aa-91da59d48b91.png">

### Expected Behavior

We expected to see a similar source map after running the migration. It was unexpected to see unused code show up in the bundle.

### Github Repo

https://github.com/smithpb/nx-tree-shaking

### Steps to Reproduce

1. Build "development" on main `nx run angular-15-migration:build:development`
2. Generate source map. I used `npx source-map-explorer dist/apps/angular-15-migration/**/*.js`
3. Checkout `migration` branch
4. `npm install`
5. Repeat steps 1 and 2 for this branch.

### Nx Report

```shell
Node : 16.14.0
   OS   : darwin x64
   npm  : 8.3.1
   
   nx : 15.5.1
   @nrwl/angular : 15.5.1
   @nrwl/cypress : 15.5.1
   @nrwl/detox : Not Found
   @nrwl/devkit : 15.5.1
   @nrwl/esbuild : Not Found
   @nrwl/eslint-plugin-nx : 15.5.1
   @nrwl/expo : Not Found
   @nrwl/express : Not Found
   @nrwl/jest : 15.5.1
   @nrwl/js : 15.5.1
   @nrwl/linter : 15.5.1
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : 15.5.1
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/rollup : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 15.5.1
   @nrwl/web : Not Found
   @nrwl/webpack : 15.5.1
   @nrwl/workspace : 15.5.1
   @nrwl/vite : Not Found
   typescript : 4.8.4
   ---------------------------------------
   Local workspace plugins:
         @cpxe/nx-plugins/custom-build
   ---------------------------------------
   Community plugins:
         @ionic/angular: 6.4.1
         @ngrx/component-store: 15.0.0
         @ngrx/effects: 15.0.0
         @ngrx/entity: 15.0.0
         @ngrx/router-store: 15.0.0
         @ngrx/store: 15.0.0
         ngx-pendo: 1.10.0
         @compodoc/compodoc: 1.1.19
         @ngrx/schematics: 15.0.0
         @ngrx/store-devtools: 15.0.0
         @nxext/capacitor: 15.0.1
         @nxext/ionic-angular: 14.0.0
         @storybook/angular: 6.5.15
         @testing-library/angular: 10.11.1
         eslint-plugin-ngrx: 2.1.2
         ng-mocks: 14.3.0
```

### Additional Information

The demo repo migrated to Nx 15.5.2, but we noticed the issue when trying the migration to 15.4.0.

**Comment by alex-okrushko** [comment]
@Coly010 here's one of our blockers

**Comment by wolfhoundjesse** [comment]
This is happening in my React project, too. I started fresh with 15.5.3 and then migrated to 15.6.1.

**Comment by Coly010** [comment]
@smithb @alex-okrushko I did a thorough investigation into this and the problem does not sit with Nx.

It's an issue introduced by Angular in V15. 

I opened an issue on their GH Repo: https://github.com/angular/angular-cli/issues/24656

I have a workaround, if it works for you:

In your `apps/<your_app_name>/tsconfig.app.json`, set `"target": "es2020"`.

The `@angular-devkit/build-angular:browser` builder will log a warning, but the service will be tree shaken correctly from your bundle. The warning may be correct, and I haven't investigated if it produces unwarranted behaviour at runtime.

**Comment by alex-okrushko** [comment]
Thanks for the update @Coly010 ! And thanks for the suggested workaround. 👍

**Comment by smithpb** [comment]
Thank you @Coly010 for taking the time to look into this issue.

**Comment by Jasonkoolman** [comment]
Tree shaking does not work at all on a freshly installed nx workspace using Webpack with Babel.

### Reproduce issue

1. Create workspace by running `npx create-nx-workspace@latest monorepo`
2. Select integrated repo > app boilerplate
3. Create React application by running `npx nx generate @nrwl/react:application shell --bundler=webpack --compiler=babel --unitTestRunner=jest --e2eTestRunner=none`

Import a library which exports a barrel file. I'll use `date-fns` as an example. Open the `App` component of the newly generated React application and import a function from `date-fns`:

```
import { getDate } from 'date-fns';

console.log(getDate); // use it somewhere

export function App() {
  return null;
}
```

Now let's look at both the development and production builds in `dist/apps/shell/main[.id].js` build and see what happened.

**Development build**
- The build is huge (~1.8MB vendor)
- All `date-fns` functions are imported. Notice the import in the main bundle:
 ```
// EXTERNAL MODULE: ../../node_modules/date-fns/index.js
var date_fns = __webpack_require__(3753);
```

**Production build**
- The build is huge (~900KiB)
- All `date-fns` functions are imported
- Code doesn't seem to be minified (i.e. comments, original function names, identations and newlines are still present)

You should see something like:
![image](https://user-images.githubusercontent.com/422729/217798417-6897df97-b733-4d78-aed2-faedcdb6646e.png)

### The fix

I die some extensive digging and managed to solve the issue by extending/overriding the default webpack config:

```js
// Fix 1: resolves tree-shaking issue
// The default is [ 'browser', 'main', 'module' ]. Thus, 'main' had preference over 'module' when Webpack reads the `package.json` files, which is not what we want. Module should become before main - the order matters!
// See https://webpack.js.org/configuration/resolve/#resolvemainfields
config.resolve.mainFields = ['module', 'main', 'browser']; 

// Fix 2: resolves minification issue by adding Terser. Terser is also capable of eliminating dead code.
// TerserJS is the Webpack 5 default minifier but we have to specify it explicitly as soon as we include more minifiers
config.optimization.minimizer.unshift(new TerserJSPlugin());
```

Full code of `webpack.config.js`:

```js
const { composePlugins, withNx } = require('@nrwl/webpack');
const { withReact } = require('@nrwl/react');
const TerserJSPlugin = require('terser-webpack-plugin');

module.exports = composePlugins(withNx(), withReact(), (config) => {
  config.resolve.mainFields = ['browser', 'module', 'main'];
  config.optimization.minimizer.unshift(new TerserJSPlugin());
  return config;
});
```

**Comment by Coly010** [comment]
@Jasonkoolman  this is a slightly different issue to the one that was originally reported.

I'm cautious about mixing two separate problems into one issue.

Can you create a new issue and link it to me, I'll do the rest of the triage from there.

**Comment by Coly010** [comment]
Angular has fixed this in `@angular-devkit/build-angular@15.1.5` Feel free to update you angular-devkit and angular/cli packages to this version and it will resolve the issue. :)
